src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.DisposeDataFlowOperationVisitor.cs

```diff
@@ -0,0 +1,302 @@
+            private readonly INamedTypeSymbol _iCollection;
+            private readonly INamedTypeSymbol _genericICollection;
+            private readonly ImmutableHashSet<INamedTypeSymbol> _disposeOwnershipTransferLikelyTypes;
+
+            public DisposeDataFlowOperationVisitor(
+                INamedTypeSymbol iDisposable,
+                INamedTypeSymbol iCollection,
+                INamedTypeSymbol genericICollection,
+                ImmutableHashSet<INamedTypeSymbol> disposeOwnershipTransferLikelyTypes,
+                DisposeAbstractValueDomain valueDomain,
+                INamedTypeSymbol containingTypeSymbol,
+                DataFlowAnalysisResult<PointsToBlockAnalysisResult, PointsToAbstractValue> pointsToAnalysisResult,
+                DataFlowAnalysisResult<NullBlockAnalysisResult, NullAbstractValue> nullAnalysisResultOpt)
+                : base(valueDomain, containingTypeSymbol, nullAnalysisResultOpt: nullAnalysisResultOpt, pointsToAnalysisResultOpt: pointsToAnalysisResult)
+            {
+                Debug.Assert(iDisposable != null);
+                Debug.Assert(disposeOwnershipTransferLikelyTypes != null);
+                Debug.Assert(pointsToAnalysisResult != null);
+
+                _iDisposable = iDisposable;
+                _iCollection = iCollection;
+                _genericICollection = genericICollection;
+                _disposeOwnershipTransferLikelyTypes = disposeOwnershipTransferLikelyTypes;
+            }
+
+            protected override DisposeAbstractValue GetAbstractDefaultValue(ITypeSymbol type) => DisposeAbstractValue.NotDisposable;
+
+            protected override DisposeAbstractValue GetAbstractValue(AbstractLocation location) => CurrentAnalysisData.TryGetValue(location, out var value) ? value : ValueDomain.UnknownOrMayBeValue;
+
+            protected override void SetAbstractValue(AbstractLocation location, DisposeAbstractValue value)
+            {
+                Debug.Assert(location.LocationType.IsDisposable(_iDisposable));
+
+                CurrentAnalysisData[location] = value;
+            }
+
+            protected override void SetAbstractValue(PointsToAbstractValue instanceLocation, DisposeAbstractValue value)
+            {
+                foreach (var location in instanceLocation.Locations)
+                {
+                    SetAbstractValue(location, value);
+                }
+            }
+
+            protected override void ResetCurrentAnalysisData(DisposeAnalysisData newAnalysisDataOpt = null)
+            {
+                // Reset the current analysis data, while ensuring that we don't violate the monotonicity, i.e. we cannot remove any existing key from currentAnalysisData.
+                if (newAnalysisDataOpt == null)
+                {
+                    // Just set the values for existing keys to ValueDomain.UnknownOrMayBeValue.
+                    var keys = CurrentAnalysisData.Keys.ToImmutableArray();
```

Comment:
I assume that we have some guarantees here of no multithreading? Otherwise, we have potential races here.

Reply:
Yes, the DataFlowOperationVisitor is always executed on a single thread, and each analysis invocation creates a new instance. The results from analysis are cached as immutable objects, which can be accessed concurrently, and computed on demand if their is a cache miss.